src/opam_file.ml

```diff
@@ -80,73 +80,25 @@ let absolutify_positions ~file_contents t =
 
 let nopos : OpamParserTypes.pos = ("",0,0) (* Null position *)
 
-module Mutator = struct
-  open OpamParserTypes
-
-  type t = opamfile_item list -> opamfile_item list
-
-  let (>>>) : t -> t -> t = fun x y z -> y (x z)
-
-  let fixup : t = List.map ~f:(function
-    | Variable (x,y,String (pos,z)) ->
-      let fixed =
-        if String.length z > 0 && z.[0] = '\n'
-        then String.sub z ~pos:1 ~len:(String.length z - 1)
-        else z
-      in
-      Variable (x,y,String (pos,fixed))
-    | y -> y)
-
-  let _remove_var : string -> t =
-    fun str -> List.filter ~f:(function
-      | Variable (_, v, _) when v=str -> false
-      | _ -> true)
-
-  let add_var : string -> OpamParserTypes.value -> t = fun var value l ->
-    (Variable (nopos, var, value))::l
-
-  let remap x f =
-    List.filter_map ~f:(function
-      | Variable (_, v, y) when v = x -> begin
-        match f (Some y) with
-        | Some y' -> Some (Variable (nopos, v, y'))
-        | None -> None
-        end
-      | z -> Some z)
-
-  let binding_present x =
-    List.exists ~f:(function
-      | Variable (_, v, _) when v = x -> true
-      | _ -> false)
-
-  let _map_var x f zs =
-    if binding_present x zs
-    then remap x f zs
-    else begin
-      match f None with
-      | Some y -> (Variable (nopos, x, y))::zs
-      | None -> zs
-    end
-
-  let set_var x y zs =
-    if binding_present x zs
-    then remap x (fun _ -> Some y) zs
-    else add_var x y zs
-
-  let mkstring x = String (nopos, x)
-  let mklist f xs = List (nopos, List.map ~f xs)
-
-  let set_string x y = set_var x (mkstring y)
-
-  let set_list x conv l = set_var x (mklist conv l)
-  let id x = x
-
-  let opt opt f : t = match opt with | None -> id | Some x -> f x
-  let list l f : t = match l with [] -> id | xs -> f xs
-
-  let apply t opamfile =
-    {
-      opamfile with
-      file_contents = t opamfile.file_contents
-    }
+module Create = struct
+  let string s = String (nopos, s)
+  let list f xs = List (nopos, List.map ~f xs)
+  let string_list xs = list string xs
+
+  let add_missing_bindings (t : t) vars =
+    let existing_variables =
+      List.fold_left ~init:String.Set.empty t.file_contents
+        ~f:(fun acc l ->
+          match l with
+          | Section (_, _) -> acc
+          | Variable (_, var, _) -> String.Set.add acc var)
+    in
+    let vars =
+      List.filter_map vars ~f:(fun (var, value) ->
+        if String.Set.mem existing_variables var then
+          None
+        else
+          Some (Variable (nopos, var, value)))
+    in
+    { t with file_contents = vars @ t.file_contents }
```

Comment:
That's what's causing the `build:` instruction to come first. What about reversing the order of argument of `@`? We could also only filter out variables here, and append the original template string unmodified to the output.

Reply:
Anil asked the fields from the template to come first.

We could do it via string concatenation, but I don't think it buys us anything. The current order is chosen consciously but I'm happy to tweak it.

Reply:
Ah, I thought this was the opposite. It doesn't bother anyone that `build:` is at the beginning of the file? That looks odd to me. We could also do the same as for `META` files: let the user choose where the generated contents is inserted.

Reply:
Cross wired I think @rgrinberg -- I intended to request the template fields come last, so I think we're all in agreement there.

https://github.com/ocaml/dune/pull/2091#issuecomment-489022808
> Also, this is aesthetic, but could the template rules be added to the bottom of the generated file? Convention is that opam-version is at the top.

Reply:
Argh! I realised the source of the confusion -- by "template fields" I'm referring to the hand-written opam.template file, whereas I think it actually means "the autogenerated fields from the builtin template".  To be clear: I mean that the `build` instruction from the override should come last :)

Reply:
Okay, I reversed the order. Sorry for the confusion.